api/src/main/java/net/starype/quiz/api/parser/TrackedDatabase.java

```diff
@@ -0,0 +1,177 @@
+
+                // Read the require number of bytes and parse the entry
+                byte[] entryRaw = new byte[nbBytes];
+                buffer.get(entryRaw);
+                DBEntry entry = new DBEntry(table);
+                entry.evaluate(ByteBuffer.wrap(entryRaw));
+                entries.add(entry);
+            }
+        }
+        catch (RuntimeException ignored) {}
+
+        // Secondly compare each file in the list with the Database
+        // Retrieve a list of file to sync
+        Set<String> syncRequired = trackedFiles.stream()
+                .filter(this::doesRequireSync)
+                .collect(Collectors.toSet());
+
+        // Discard all the file no longer tracked
+        entries = entries
+                .stream()
+                .filter(entry -> trackedFiles.stream().anyMatch(trackedFile -> trackedFile.equals(entry.file())))
+                .collect(Collectors.toSet());
+
+        // Sync each file
+        syncRequired.forEach(this::sync);
+        syncOut();
+    }
+
+    private void sync(String file) {
+        // Compute the checksum of the file
+        CheckSum checkSum = CheckSum.fromFile(file).orElseThrow();
+
+        // Update the entries (remove all the old instance of the file)
+        entries = entries.stream()
+                .filter(entry -> !entry.file().equals(file))
+                .collect(Collectors.toSet());
+
+        // Adding the entries present in the file to the DB
+        Set<DBEntry> newEntries = parser.read(file);
+        newEntries.forEach(entry -> {
+            entry.setFile(file);
+            entry.setCheckSum(checkSum);
+        });
+        entries.addAll(newEntries);
+    }
+
+    private void syncOut() {
+        try {
+            // First create a new output stream
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+
+            // Secondly output the number of file
+            outputStream.write(ByteBuffer.allocate(4).putInt(entries.size()).array());
+
+            // Then output each file individually
+            for(DBEntry entry : entries) {
+                // Compile the data of the entry
+                ByteBuffer buffer = entry.save();
+
+                // Write the ByteBuffer data
+                outputStream.write(ByteBuffer.allocate(4).putInt(buffer.array().length).array());
+
+                // Write the buffer to the stream
+                outputStream.write(buffer.array());
+            }
+
+            io.write(ByteBuffer.wrap(outputStream.toByteArray()));
+        }
+        catch(IOException ignored) {}
+    }
+
+    private boolean doesRequireSync(String file) {
+        return entries.stream()
+                .noneMatch(entry -> entry.file().equals(file)) ||
+               entries.stream()
+                       .filter(entry -> entry.file().equals(file))
+                       .noneMatch(entry -> entry.checkSum().equals(CheckSum.fromFile(file).orElse(CheckSum.NIL)));
```

Comment:
Can't you use de Morgan's laws to have only one stream ? 
`(∀¬P(x)) v (∀¬S(x)) = ¬ (∃P(x) ^ ∃S(x))`

Reply:
Well we will still have two stream because we cannot rewrite that as
![image](https://user-images.githubusercontent.com/49918742/105535311-083bad80-5cef-11eb-81e7-9e3af0b5c05e.png)